src/release/check.rs

```diff
@@ -0,0 +1,117 @@
+use crate::release_api::{ApiError, Release};
+use ubuntu_version::{Version, VersionError};
+
+pub enum BuildStatus {
+    Blacklisted,
+    Build(u16),
+    ConnectionIssue(reqwest::Error),
+    InternalIssue(ApiError),
+    ServerStatus(reqwest::Error),
+}
+
+impl BuildStatus {
+    pub fn is_ok(&self) -> bool {
+        if let BuildStatus::Build(_) = *self {
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn status_code(&self) -> i16 {
+        match *self {
+            BuildStatus::ConnectionIssue(_) => -3,
+            BuildStatus::ServerStatus(_) => -2,
+            BuildStatus::InternalIssue(_) => -1,
+            BuildStatus::Build(build) => build as i16,
+            BuildStatus::Blacklisted => -4,
+        }
+    }
+}
+
+impl From<Result<u16, ApiError>> for BuildStatus {
+    fn from(result: Result<u16, ApiError>) -> Self {
+        match result {
+            Err(ApiError::Get(why)) => BuildStatus::ConnectionIssue(why),
+            Err(ApiError::Status(why)) => BuildStatus::ServerStatus(why),
+            Err(otherwise) => BuildStatus::InternalIssue(otherwise),
+            Ok(build) => BuildStatus::Build(build),
+        }
+    }
+}
+
+pub struct ReleaseStatus {
+    pub current: &'static str,
+    pub next:    &'static str,
+    pub build:   BuildStatus,
+    pub is_lts:  bool,
+}
+
+impl ReleaseStatus {
+    pub fn is_lts(&self) -> bool { self.is_lts }
+}
+
+pub fn next(development: bool) -> Result<ReleaseStatus, VersionError> {
+    let current = Version::detect()?;
+
+    let next: &str;
+
+    let status = match (current.major, current.minor) {
+        (18, 4) => {
+            next = if development { "20.04" } else { "19.04" };
+
+            ReleaseStatus {
+                build: Release::build_exists(next, "intel").into(),
+                current: "18.04",
+                is_lts: true,
+                next,
+            }
+        }
+
+        (19, 10) => {
+            next = "20.04";
+
+            ReleaseStatus {
+                build: if development {
+                    Release::build_exists(next, "intel").into()
+                } else {
+                    BuildStatus::Blacklisted
+                },
+                current: "19.10",
+                is_lts: false,
+                next,
+            }
+        }
+
+        (20, 4) => ReleaseStatus {
+            build:   BuildStatus::Blacklisted,
+            current: "20.04",
+            is_lts:  true,
+            next:    "20.10",
```

Comment:
Wouldn't this allow 20.04 to upgrade to 20.10 immediately?

Reply:
Not unless `build` contains a `BuildStatus::Build` signifying that an ISO is available.

Reply:
That is going to happen eventually so shouldn't it also do what is done in the 19.10 listing?